### PR TITLE
Fix scenario service vs struct projections bug

### PR DESCRIPTION
**Pull request by hurryabit**
In a recent change, we have sped up the projection from structural
records by means of an inference in the DAML-LF type checker. This
change was made under the assumption that users of the DAML interpreter
run through package validation before compiling to the internal Speedy
AST. Unfortunately, that assumption was wrong and not covered by our
existing tests.  Manually testing of a new release candidate reminded us
of the fact that the scenario service skips package validation for the
sake of faster response times in the IDE.

This PR drops the assumption that package validation is always run.
Instead, we add the old implementation of struct projection, which works
without DAML-LF type checking, back and use it whenever the AST has not
been annotated with the information inferred by the type checker. We do
the same for struct updates.

We improve test coverage by _additionally_ running the `damlc`
integration tests without package validation for the latest stable
version of DAML-LF and DAML-LF 1.dev. These tests would have caught the
issue we only discovered during manual testing.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

---
This change is [Reviewable](https://reviewable.io/reviews/digital-asset/daml/7778)

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 1]
```diff
@@ -797,7 +797,19 @@ private[lf] object SBuiltin {
     }
   }
 
-  /** $tupd[field] :: Struct -> a -> Struct */
+  /** $tproj[field] :: Struct -> a */
+  final case class SBStructProjByName(field: Ast.FieldName) extends SBuiltinPure(1) {
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:801]: Why not adding a comment to explicate that this is a non-optimize form of `SBStructProj`

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 2]
```diff
@@ -376,16 +376,13 @@ private[lf] final class Compiler(
         )
       case structProj: EStructProj =>
         structProj.fieldIndex match {
-          case None =>
-            throw CompilationError(
-              s"structural record projection for field ${structProj.field} has no index")
+          case None => SBStructProjByName(structProj.field)(compile(structProj.struct))
           case Some(index) => SBStructProj(index)(compile(structProj.struct))
         }
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:381]: Why not something like: 

```scala
   val builtin = structProj.fieldIndex.fold(SBStructProjByName(structProj.field))(SBStructProj)
   builtin.apply(compile(structProj.struct))
```

**Review by remyhaemmerle-da — APPROVED** [review]
Looks reasonable.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala (hunk 2 above)
Comment (hurryabit) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:381]: I simply don't like `fold` on `Option`. Whenever I see `fold`, I need to stop for a moment and run a type inference in my head. I don't need to do this with pattern matching. Thus, I'm avoiding to produce code I _personally_ find harder to read, although I won't stop others from using `fold`.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 1 above)
Comment (hurryabit) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:801]: Good idea. I'll do that.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:381]: No problem, it was just a suggestion.

Comment (hurryabit) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:381]: I know. Apologies if I came across as aggressive!

My goal was to share my motivation for doing things the way I do them. 😃

**Review by cocreature — APPROVED** [review]
Thanks!

**Comment by nickchapman-da** [comment]
When this PR lands in master, I will cut a new snapshot release.

**Review by nickchapman-da — APPROVED** [review]

**Pull request merged**